src/grasplan/tools/action_client_helper.py

```diff
@@ -0,0 +1,209 @@
+            self.goal_set.add(goal_id)
+            rospy.loginfo(f'Captured new goal_id for {self.canceller_type}: {goal_id}')
+
+    def status_callback(self, msg):
+        # Iterate over each status in the status_list
+        for status in msg.status_list:
+            goal_id = status.goal_id.id
+            goal_status = status.status
+            # If the status is SUCCEEDED, remove it from the set
+            if goal_status == GoalStatus.SUCCEEDED:
+                if goal_id in self.goal_set:
+                    self.goal_set.remove(goal_id)
+                    rospy.loginfo(f'Goal {goal_id} succeeded and removed from set ({self.canceller_type})')
+
+    def has_active_goals(self):
+        # Check if there are any goals in the set
+        return len(self.goal_set) > 0
+
+    def cancel_all_goals(self):
+        if self.has_active_goals():
+            # Iterate over all goal_ids in the set, sending a cancel request for each
+            while self.goal_set:
+                goal_id = self.goal_set.pop()  # Remove any goal_id from the set
+                cancel_msg = GoalID()
+                cancel_msg.id = goal_id
+                self.cancel_pub.publish(cancel_msg)
+                rospy.loginfo(f'Sent {self.canceller_type} cancel request for goal_id: {goal_id}')
+            return True
+        else:
+            rospy.logwarn(f'No active goals to cancel for {self.canceller_type}')
+            return False
+
+
+class ActionClientHelper:
+    """
+    A customized alternative to action_client.wait_for_result(...)
+    that includes support for bypassing unresponsive action servers (e.g., MoveIt pickup server).
+    If the server ignores preemption requests, this method forcibly cancels goals
+    by directly interacting with the joint trajectory controllers.
+    """
+
+    def __init__(self, ns, action_server, controller_names):
+        """
+        Initializes the action server helper.
+
+        :param ns: The namespace under which the action server and controllers are operating.
+        :param type: A string indicating the type of action server, e.g., 'pick', 'place', or 'insert'.
+        :param action_server: An instance of the actionlib server that manages the corresponding actions.
+        :param controller_names: A list of controller names, e.g., ['arm', 'gripper'],
+                                used to create trajectory cancellers for each controller.
+
+        Example:
+            If `ns` is '/robot_namespace' and `controller_names` is ['arm', 'gripper'],
+            the trajectory canceller topics will be:
+                - /robot_namespace/arm_controller/follow_joint_trajectory/cancel
+                - /robot_namespace/gripper_controller/follow_joint_trajectory/cancel
+        """
+        self.canceller_list = []
+        for controller_name in controller_names:
+            self.canceller_list.append(
+                TrajectoryCanceller(f'/{ns}/{controller_name}_controller/follow_joint_trajectory/')
```

Comment:
NOTE: two object instances are created, one for the arm and one for the gripper, however note that this is done 1 time for pick, 1 time for place and 1 for insert, making in total 6 subscribers. This is not ideal... but unfortunately an optimal solution of having 2 cancellers (one for arm and one for gripper) that is shared between the pick, place and insert modules, would require, option 1) heavy refactoring of Grasplan code or 2) having the canceller as a separate node and making service calls to it. For now I leave it like this, but we can have a more in-depth discussion about it.